### ISSUE TITLE
TestRegistryClientImageV1/TestRegistryClientImageV2 failure

### ISSUE THREAD

**Issue by kargakis** [openshift/origin#9130]
```
--- FAIL: TestRegistryClientImageV1 (1.46s)
	dockerregistryclient_test.go:222: the default tag "sha256:52776e94e88f0335b7165113ea40ebdf0c919eb208d357287537544987d14b19" has not been set on repository "openshift/origin"
FAIL
```
Same error for `TestRegistryClientImageV2`
https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_integration/1423/console

**Comment by kargakis** [comment]
Bumping to P0 because it seems it's not a flake but a real failure

**Comment by legionus** [comment]
I can tell you more:
```
--- FAIL: TestRegistryClientImageV2 (5.27s)
	dockerregistryclient_test.go:222: the default tag "sha256:52776e94e88f0335b7165113ea40ebdf0c919eb208d357287537544987d14b19" has not been set on repository "openshift/origin"
```

`TestRegistryClientImageV2` flakes too.

**Comment by legionus** [comment]
full log:
```
=== RUN   TestRegistryClientImageV2
I0602 11:54:49.069659   25787 client.go:320] Found registry v2 API at https://registry-1.docker.io/v2/
I0602 11:54:52.695341   25787 client.go:555] passed valid auth token, but unable to find tagged image at "https://registry-1.docker.io/v2/openshift/origin-not-found/manifests/latest", 401 map[Content-Type:[application/json; charset=utf-8] Docker-Distribution-Api-Version:[registry/2.0] Www-Authenticate:[Bearer realm="https://auth.docker.io/token",service="registry.docker.io",scope="repository:openshift/origin-not-found:pull",error="insufficient_scope"] Date:[Thu, 02 Jun 2016 09:54:52 GMT] Content-Length:[158] Strict-Transport-Security:[max-age=31536000]]: {"errors":[{"code":"UNAUTHORIZED","message":"authentication required","detail":[{"Type":"repository","Name":"openshift/origin-not-found","Action":"pull"}]}]}
I0602 11:54:53.796195   25787 client.go:567] unable to find tagged image at "https://registry-1.docker.io/v2/openshift/origin/manifests/sha256:52776e94e88f0335b7165113ea40ebdf0c919eb208d357287537544987d14b19", 404 map[Content-Type:[application/json; charset=utf-8] Docker-Distribution-Api-Version:[registry/2.0] Date:[Thu, 02 Jun 2016 09:54:53 GMT] Content-Length:[192] Strict-Transport-Security:[max-age=31536000]]: {"errors":[{"code":"MANIFEST_UNKNOWN","message":"manifest unknown","detail":{"Name":"openshift/origin","Revision":"sha256:52776e94e88f0335b7165113ea40ebdf0c919eb208d357287537544987d14b19"}}]}
--- FAIL: TestRegistryClientImageV2 (5.27s)
	dockerregistryclient_test.go:222: the default tag "sha256:52776e94e88f0335b7165113ea40ebdf0c919eb208d357287537544987d14b19" has not been set on repository "openshift/origin"
FAIL
```

**Comment by legionus** [comment]
and more:
```
=== RUN   TestImageStreamImportDockerHub
I0602 11:50:28.690466   23916 importer.go:305] importing remote Docker repository registry=https://registry-1.docker.io repository=library/redis insecure=false
I0602 11:50:29.232394   23916 credentials.go:30] asked to provide RefreshToken for https://auth.docker.io/token
I0602 11:50:29.232411   23916 credentials.go:25] asked to provide Basic credentials for https://auth.docker.io/token
I0602 11:50:30.508498   23916 importer.go:305] importing remote Docker repository registry=https://registry-1.docker.io repository=library/mysql insecure=false
I0602 11:50:30.509669   23916 credentials.go:30] asked to provide RefreshToken for https://auth.docker.io/token
I0602 11:50:30.509680   23916 credentials.go:25] asked to provide Basic credentials for https://auth.docker.io/token
I0602 11:50:31.010503   23916 importer.go:305] importing remote Docker repository registry=https://registry-1.docker.io repository=mysql/doesnotexistinanyform insecure=false
I0602 11:50:31.011715   23916 credentials.go:30] asked to provide RefreshToken for https://auth.docker.io/token
I0602 11:50:31.011736   23916 credentials.go:25] asked to provide Basic credentials for https://auth.docker.io/token
I0602 11:50:31.269478   23916 client.go:326] Retrying request to a v2 Docker registry after encountering error (1 attempts remaining): unauthorized: authentication required
I0602 11:50:32.400343   23916 client.go:326] Retrying request to a v2 Docker registry after encountering error (0 attempts remaining): unauthorized: authentication required
I0602 11:50:32.531135   23916 importer.go:437] unable to get tag "latest" for repository &importer.importRepository{Ref:api.DockerImageReference{Registry:"", Namespace:"mysql", Name:"doesnotexistinanyform", Tag:"", ID:""}, Registry:(*url.URL)(0xc820348900), Name:"mysql/doesnotexistinanyform", Insecure:false, Tags:[]importer.importTag{importer.importTag{Name:"latest", Image:(*api.Image)(nil), Err:error(nil)}}, Digests:[]importer.importDigest(nil), MaximumTags:0, AdditionalTags:[]string(nil), Err:error(nil)}: errcode.Error{Code:1002, Message:"authentication required", Detail:[]uint8{}}
I0602 11:50:32.531356   23916 importer.go:305] importing remote Docker repository registry=https://registry-1.docker.io repository=library/mongo insecure=false
I0602 11:50:32.535871   23916 credentials.go:30] asked to provide RefreshToken for https://auth.docker.io/token
I0602 11:50:32.535915   23916 credentials.go:25] asked to provide Basic credentials for https://auth.docker.io/token
--- FAIL: TestImageStreamImportDockerHub (5.99s)
	imageimporter_test.go:696: unexpected object: api.ImageImportStatus{Tag:"latest", Status:unversioned.Status{TypeMeta:unversioned.TypeMeta{Kind:"", APIVersion:""}, ListMeta:unversioned.ListMeta{SelfLink:"", ResourceVersion:""}, Status:"Failure", Message:"Internal error occurred: unauthorized: authentication required", Reason:"InternalError", Details:(*unversioned.StatusDetails)(0xc82072eb40), Code:500}, Image:(*api.Image)(nil)}
FAIL
```

**Comment by kargakis** [comment]
cc: @soltysh @smarterclayton @miminar @deads2k @liggitt @mfojtik

**Comment by kargakis** [comment]
@legionus is `TestImageStreamImportDockerHub` from your 2.4 branch? I cannot see it on master.

**Comment by legionus** [comment]
@kargakis https://github.com/openshift/origin/blob/master/test/integration/imageimporter_test.go#L625

**Comment by legionus** [comment]
add for the company:
```
Running test/cmd/newapp.sh:76: executing 'oc new-app --search ruby-helloworld-sample' expecting success and text 'ruby-helloworld-sample'...
FAILURE after 1.681s: test/cmd/newapp.sh:76: executing 'oc new-app --search ruby-helloworld-sample' expecting success and text 'ruby-helloworld-sample': the command returned the wrong error code
There was no output from the command.
Standard error from the command:
error: can't look up Docker image "ruby-helloworld-sample": Internal error occurred: unauthorized: authentication required
```

**Comment by kargakis** [comment]
> @kargakis https://github.com/openshift/origin/blob/master/test/integration/imageimporter_test.go#L625

sorry, I was referring to the failure you posted, not the test per se

**Comment by legionus** [comment]
Yes, I was wrong.

**Comment by legionus** [comment]
It seems we get an invalid token.

**Comment by smarterclayton** [comment]
The failure is because the digest returned is for schema2, not schema1, and
so the fetch fails.

**Comment by smarterclayton** [comment]
Or at least, the manifest that has been recorded as being "latest" is not
the actual manifest - but fetching the content doesn't return the actual
content.

**Comment by smarterclayton** [comment]
So the test is unfixable at this point.  We need to test against another v2 registry that has the old behavior (quay) and then simply report this as "expected" because the content type is schema1.

**Comment by smarterclayton** [comment]
We also need to update to be building and pushing images to the hub with Docker 1.10.

**Comment by smarterclayton** [comment]
@danmcp what's the latest on switching the CI machines to Docker 1.10?

**Comment by danmcp** [comment]
@smarterclayton What ci systems?  We are at 1.10 with all the rhel (the default ones) instances afaik.

**Comment by smarterclayton** [comment]
So we've had a clean devenv AMI that has used docker 1.10 to push to the hub?

**Comment by legionus** [comment]
@smarterclayton I added few debug messages to `dockerregistry/client`. It's not looks like a problem with the scheme v2.

https://gist.github.com/legionus/41f419a9348113da3a644ef87275e285

**Comment by smarterclayton** [comment]
It looks like it - https://ci.openshift.redhat.com/jenkins/job/devenv_ami/4302/console pushed sha256:95067af9d8eba9248c663dced2227477d7d367c475345b8513fec056b2366806

**Comment by smarterclayton** [comment]
When the content type for the manifest fetch is `application/vnd.docker.distribution.manifest.v2+json` then the ID can be fetched (so it's definitely that we're getting the wrong content digest).  So the question is how we deal with older clients - should we always try schema2 first, then try schema1?

**Comment by smarterclayton** [comment]
quay.io responds with schema1 even if we send the schema2 manifest still.  We should test what happens on our registry if we ask for 2 then 1 for an image pushed as 1.

**Comment by smarterclayton** [comment]
It looks like the current registry (2.2 and below) will correctly respond for:

```
Accept: application/vnd.docker.distribution.manifest.v2+json,application/vnd.docker.distribution.manifest.v1+json
```

So I think our legacy client can be updated to perform that action, as long as we correctly handle the v2 schema.  That will fix this failure mode for any client on the newest code.  Old clients will still be broken, which raises the question of whether we patch this on 1.2.0.

**Comment by smarterclayton** [comment]
Tests are disabled, dropping this to p1.

**Comment by smarterclayton** [comment]
To summarize:

1. DockerHub is broken on schema1 for pull by digest (known)
2. Once we pushed a new image to Docker Hub for openshift/origin:latest from Docker 1.10, we broke ourselves (our test case depends on origin:latest resolving via digest)
3. We may need to add schema2 support to the legacy dockerregistry client - we still require it to support importing from old registries, so it's not a high priority
4. We should add another test case for a schema1 image on the dockerhub that isn't origin:latest to verify the old code path still works
5. We didn't react well to this change - we should probably do a post-mortem on the platform management team side to discuss what we could do better in the future.
